File: platform/sdk/go/native_capabilities_test.go
package layerx

import (
	"bytes"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"os"
	"testing"
)

type nativeCapabilityFixtureGrant struct {
	Tag           byte   `json:"tag"`
	Kind          string `json:"kind"`
	Program       string `json:"program"`
	OwnerProgram  string `json:"owner_program"`
	Seed          string `json:"seed"`
	SourceAccount string `json:"source_account"`
	Asset         string `json:"asset"`
	To            string `json:"to"`
	Account       string `json:"account"`
	ReceiptDigest string `json:"receipt_digest"`
	MaximumAmount string `json:"maximum_amount"`
}

func nativeCapabilityFixtureGrants(t *testing.T, values []nativeCapabilityFixtureGrant) []NativeCapability {
	t.Helper()
	field := func(value string) [32]byte {
		decoded, err := hex.DecodeString(value)
		if err != nil || len(decoded) != 32 {
			t.Fatalf("invalid identifier %q", value)
		}
		var result [32]byte
		copy(result[:], decoded)
		return result
	}
	amount := func(value string) Uint128 {
		parsed, err := ParseUint128(value)
		if err != nil {
			t.Fatal(err)
		}
		return parsed
	}
	grants := make([]NativeCapability, 0, len(values))
	for _, value := range values {
		var grant NativeCapability
		switch value.Kind {
		case "StorageRead":
			grant = NativeStorageRead{}
		case "StorageWrite":
			grant = NativeStorageWrite{}
		case "EmitEvent":
			grant = NativeEmitEvent{}
		case "Call":
			grant = NativeCallCapability{field(value.Program)}
		case "Transfer402":
			grant = NativeTransfer402{field(value.Asset), field(value.To), amount(value.MaximumAmount)}
		case "ProgramSpend":
			seed, err := hex.DecodeString(value.Seed)
			if err != nil {
				t.Fatal(err)
			}
			grant = NativeProgramSpend{field(value.OwnerProgram), seed, field(value.SourceAccount), field(value.Asset), field(value.To), amount(value.MaximumAmount)}
		case "ReceiptRead":
			grant = NativeReceiptRead{field(value.ReceiptDigest)}
		case "BalanceView":
			grant = NativeBalanceView{field(value.Account), field(value.Asset), field(value.ReceiptDigest)}
		case "SharedStorageRead":
			grant = NativeSharedStorageRead{}
		case "SharedStorageWrite":
			grant = NativeSharedStorageWrite{}
		default:
			t.Fatalf("unknown fixture capability %q", value.Kind)
		}
		entry, err := nativeCapabilityEntryFor(grant)
		if err != nil || entry.encoded[0] != value.Tag {
			t.Fatalf("fixture tag differs: %v", err)
		}
		grants = append(grants, grant)
	}
	return grants
}

func TestNativeCapabilityRuntimeFixture(t *testing.T) {
	raw, err := os.ReadFile("../conformance/fixtures/native-program-capabilities-v2.json")
	if err != nil {
		t.Fatal(err)
	}
	var fixture struct {
		Capabilities []nativeCapabilityFixtureGrant `json:"capabilities"`
		Canonical    string                         `json:"canonical_hex"`
		Narrowed     []nativeCapabilityFixtureGrant `json:"narrowed_capabilities"`
		NarrowedHex  string                         `json:"narrowed_hex"`
		Equal        bool                           `json:"equal_narrowing_accepted"`
		Escalations  []struct {
			Name         string                         `json:"name"`
			Parent       string                         `json:"parent"`
			Capabilities []nativeCapabilityFixtureGrant `json:"capabilities"`
			Canonical    string                         `json:"canonical_hex"`
			Accepted     bool                           `json:"accepted"`
		} `json:"escalation_cases"`
	}
	if err := json.Unmarshal(raw, &fixture); err != nil {
		t.Fatal(err)
	}
	if len(fixture.Capabilities) != 10 || len(fixture.Escalations) != 3 || !fixture.Equal {
		t.Fatal("runtime coverage differs")
	}
	for index, tag := range []byte{1, 2, 3, 4, 5, 9, 6, 10, 7, 8} {
		if fixture.Capabilities[index].Tag != tag {
			t.Fatal("canonical runtime order differs")
		}
	}
	parent := nativeCapabilityFixtureGrants(t, fixture.Capabilities)
	assertEncoding := func(grants []NativeCapability, expected string) {
		t.Helper()
		encoded, err := EncodeNativeCapabilities(grants)
		if err != nil || hex.EncodeToString(encoded) != expected {
			t.Fatalf("runtime encoding differs: %v", err)
		}
		decoded, err := DecodeNativeCapabilities(encoded)
		if err != nil {
			t.Fatal(err)
		}
		reencoded, err := EncodeNativeCapabilities(decoded)
		if err != nil || !bytes.Equal(encoded, reencoded) {
			t.Fatalf("reencode differs: %v", err)
		}
	}
	assertEncoding(parent, fixture.Canonical)
	requested := nativeCapabilityFixtureGrants(t, fixture.Narrowed)
	assertEncoding(requested, fixture.NarrowedHex)
	narrowed, err := NarrowNativeCapabilities(parent, requested)
	if err != nil {
		t.Fatal(err)
	}
	assertEncoding(narrowed, fixture.NarrowedHex)
	equal, err := NarrowNativeCapabilities(parent, parent)
	if err != nil {
		t.Fatal(err)
	}
	assertEncoding(equal, fixture.Canonical)
	for _, escalation := range fixture.Escalations {
		if escalation.Parent != "narrowed" || escalation.Accepted {
			t.Fatal("runtime escalation metadata differs")
		}
		grants := nativeCapabilityFixtureGrants(t, escalation.Capabilities)
		assertEncoding(grants, escalation.Canonical)
		if _, err := NarrowNativeCapabilities(narrowed, grants); err == nil {
			t.Fatalf("escalation accepted: %s", escalation.Name)
		}
	}
}

func capabilityIdentifier(value byte) [32]byte {
	var identifier [32]byte
	for index := range identifier {
		identifier[index] = value
	}
	return identifier
}

func TestNativeCapabilityBoundsAndNarrowing(t *testing.T) {
	owner, asset, to, receipt := capabilityIdentifier(1), capabilityIdentifier(2), capabilityIdentifier(3), capabilityIdentifier(4)
	seed := []byte{0, 255}
	source, err := DeriveNativeProgramAccount(owner, seed)
	if err != nil {
		t.Fatal(err)
	}
	maximum := NewUint128(1, 0)
	spend := NativeProgramSpend{owner, seed, source, asset, to, maximum}
	view := NativeBalanceView{source, asset, receipt}
	parent := []NativeCapability{NativeSharedStorageWrite{}, NativeSharedStorageRead{}, view, NativeReceiptRead{receipt}, spend, NativeTransfer402{asset, to, maximum}, NativeCallCapability{owner}, NativeEmitEvent{}, NativeStorageWrite{}, NativeStorageRead{}}
	encoded, err := EncodeNativeCapabilities(parent)
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(encoded[:5], []byte{0, 10, 1, 2, 3}) {
		t.Fatal("wrong canonical prefix")
	}
	childSpend := spend
	childSpend.MaximumAmount = NewUint128(0, ^uint64(0))
	child, err := NarrowNativeCapabilities(parent, []NativeCapability{childSpend, view})
	if err != nil || len(child) != 2 {
		t.Fatalf("amount narrowing refused: %v", err)
	}
	childSpend.MaximumAmount = NewUint128(1, 1)
	if _, err := NarrowNativeCapabilities(parent, []NativeCapability{childSpend}); err == nil {
		t.Fatal("amount escalation accepted")
	}
	changedView := view
	changedView.ReceiptDigest = capabilityIdentifier(5)
	if _, err := NarrowNativeCapabilities(parent, []NativeCapability{changedView}); err == nil {
		t.Fatal("receipt substitution accepted")
	}
	if _, err := EncodeNativeCapabilities([]NativeCapability{view, changedView}); err == nil {
		t.Fatal("duplicate balance key accepted")
	}
	if _, err := NarrowNativeCapabilities(nil, []NativeCapability{NativeStorageRead{}}); err == nil {
		t.Fatal("missing authority accepted")
	}
	if _, err := EncodeNativeCapabilities([]NativeCapability{NativeTransfer402{asset, to, Uint128{}}}); err == nil {
		t.Fatal("zero amount accepted")
	}
	if _, err := EncodeNativeCapabilities([]NativeCapability{NativeCallCapability{}}); err == nil {
		t.Fatal("zero program accepted")
	}
	changedSpend := spend
	changedSpend.SourceAccount[0] ^= 1
	if _, err := EncodeNativeCapabilities([]NativeCapability{changedSpend}); err == nil {
		t.Fatal("forged program account accepted")
	}
	changedSpend = spend
	changedSpend.Seed = make([]byte, 129)
	if _, err := EncodeNativeCapabilities([]NativeCapability{changedSpend}); err == nil {
		t.Fatal("oversized seed accepted")
	}
	for length := 0; length < len(encoded); length++ {
		if _, err := DecodeNativeCapabilities(encoded[:length]); err == nil {
			t.Fatalf("accepted prefix %d", length)
		}
	}
	if _, err := DecodeNativeCapabilities(append(append([]byte{}, encoded...), 0)); err == nil {
		t.Fatal("accepted trailing bytes")
	}
	for _, malformed := range [][]byte{{0, 2, 2, 1}, {0, 2, 1, 1}, {0, 1, 11}, {0, 239}, {0, 1, 9}} {
		if _, err := DecodeNativeCapabilities(malformed); err == nil {
			t.Fatalf("malformed set accepted: %x", malformed)
		}
	}
	views := make([]NativeCapability, 0, 33)
	for index := byte(1); index <= 33; index++ {
		views = append(views, NativeBalanceView{capabilityIdentifier(index), asset, receipt})
	}
	if _, err := EncodeNativeCapabilities(views[:32]); err != nil {
		t.Fatal(err)
	}
	if _, err := EncodeNativeCapabilities(views); err == nil {
		t.Fatal("33 balance views accepted")
	}
	full := make([]NativeCapability, 0, MaximumNativeCapabilities+1)
	for index := 0; index <= MaximumNativeCapabilities; index++ {
		fullSeed := make([]byte, 128)
		binary.BigEndian.PutUint16(fullSeed, uint16(index))
		account, err := DeriveNativeProgramAccount(owner, fullSeed)
		if err != nil {
			t.Fatal(err)
		}
		full = append(full, NativeProgramSpend{owner, fullSeed, account, asset, to, maximum})
	}
	maximumBytes, err := EncodeNativeCapabilities(full[:MaximumNativeCapabilities])
	if err != nil || len(maximumBytes) != MaximumNativeCapabilityBytes {
		t.Fatalf("maximum canonical set: %v", err)
	}
	if _, err := EncodeNativeCapabilities(full); err == nil {
		t.Fatal("239 grants accepted")
	}
	if !validProgramCapabilities(encoded, true) || validProgramCapabilities(encoded, false) {
		t.Fatal("ABI v2 gate differs")
	}
}

File: agent/sdk/typescript/test/native-capabilities.test.ts
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import { decodeNativeCapabilitySet, encodeNativeCapabilitySet, narrowNativeCapabilitySet, deriveNativeProgramAccount, type NativeCapability } from "../src/native-capabilities.js";

type LogicalGrant = Readonly<Record<string, string | number>>;
function logical(grants: readonly LogicalGrant[]): readonly NativeCapability[] {
  const bytes = (grant: LogicalGrant, name: string) => { const value = grant[name]; assert.equal(typeof value, "string"); assert.match(String(value), /^(?:[0-9a-f]{2})*$/u); return Buffer.from(String(value), "hex"); };
  return grants.map((grant): NativeCapability => {
    switch (grant.kind) {
      case "StorageRead": return { kind: "storage_read" };
      case "StorageWrite": return { kind: "storage_write" };
      case "EmitEvent": return { kind: "emit_event" };
      case "Call": return { kind: "call", program: bytes(grant, "program") };
      case "Transfer402": return { kind: "transfer402", asset: bytes(grant, "asset"), to: bytes(grant, "to"), maximumAmount: BigInt(grant.maximum_amount!) };
      case "ProgramSpend": return { kind: "program_spend", ownerProgram: bytes(grant, "owner_program"), seed: bytes(grant, "seed"), sourceAccount: bytes(grant, "source_account"), asset: bytes(grant, "asset"), to: bytes(grant, "to"), maximumAmount: BigInt(grant.maximum_amount!) };
      case "ReceiptRead": return { kind: "receipt_read", receiptDigest: bytes(grant, "receipt_digest") };
      case "BalanceView": return { kind: "balance_view", account: bytes(grant, "account"), asset: bytes(grant, "asset"), receiptDigest: bytes(grant, "receipt_digest") };
      case "SharedStorageRead": return { kind: "shared_storage_read" };
      case "SharedStorageWrite": return { kind: "shared_storage_write" };
      default: throw new TypeError("unknown fixture grant");
    }
  });
}
const fixture = JSON.parse(readFileSync(new URL("../../../../../platform/sdk/conformance/fixtures/native-program-capabilities-v2.json", import.meta.url), "utf8")) as {
  capabilities: LogicalGrant[]; canonical_hex: string; narrowed_capabilities: LogicalGrant[]; narrowed_hex: string; equal_narrowing_accepted: boolean;
  escalation_cases: { parent: string; capabilities: LogicalGrant[]; canonical_hex: string; accepted: boolean }[];
};
const all = logical(fixture.capabilities), narrowed = logical(fixture.narrowed_capabilities);
assert.deepEqual(fixture.capabilities.map((grant) => grant.tag), [1,2,3,4,5,9,6,10,7,8]);
assert.equal(Buffer.from(await encodeNativeCapabilitySet(all)).toString("hex"), fixture.canonical_hex);
assert.equal(Buffer.from(await encodeNativeCapabilitySet(narrowed)).toString("hex"), fixture.narrowed_hex);
assert.equal(Buffer.from(await encodeNativeCapabilitySet(await narrowNativeCapabilitySet(all, narrowed))).toString("hex"), fixture.narrowed_hex);
assert.equal(fixture.equal_narrowing_accepted, true);
assert.equal(Buffer.from(await encodeNativeCapabilitySet(await narrowNativeCapabilitySet(all, all))).toString("hex"), fixture.canonical_hex);
assert.equal(fixture.escalation_cases.length, 3);
for (const entry of fixture.escalation_cases) {
  assert.equal(entry.parent, "narrowed"); assert.equal(entry.accepted, false);
  const child = logical(entry.capabilities);
  assert.equal(Buffer.from(await encodeNativeCapabilitySet(child)).toString("hex"), entry.canonical_hex);
  await assert.rejects(narrowNativeCapabilitySet(narrowed, child));
}
const encoded = Buffer.from(fixture.canonical_hex, "hex");
assert.equal(Buffer.from(await encodeNativeCapabilitySet(await decodeNativeCapabilitySet(encoded))).toString("hex"), fixture.canonical_hex);
for (let length = 0; length < encoded.length; length++) await assert.rejects(decodeNativeCapabilitySet(encoded.subarray(0, length)));
await assert.rejects(decodeNativeCapabilitySet(Buffer.concat([encoded, Buffer.from([0])])));
const unknown = Buffer.from(encoded); unknown[2] = 11; await assert.rejects(decodeNativeCapabilitySet(unknown));
const reordered = Buffer.from(encoded); reordered[2] = 2; reordered[3] = 1; await assert.rejects(decodeNativeCapabilitySet(reordered));
await assert.rejects(encodeNativeCapabilitySet([...all, all[0]!]));
const spend = all.find((grant) => grant.kind === "program_spend"); assert(spend?.kind === "program_spend");
const maximum: NativeCapability[] = [];
for (let index = 0; index < 238; index++) {
  const seed = new Uint8Array(128); new DataView(seed.buffer).setUint16(0, index);
  maximum.push({ ...spend, seed, sourceAccount: await deriveNativeProgramAccount(spend.ownerProgram, seed) });
}
assert.equal((await encodeNativeCapabilitySet(maximum)).length, 65_452);
await assert.rejects(encodeNativeCapabilitySet([...maximum, { kind: "call", program: new Uint8Array(32).fill(1) }]));
const view = all.find((grant) => grant.kind === "balance_view"); assert(view?.kind === "balance_view");
const views = Array.from({ length: 32 }, (_, index) => ({ ...view, account: new Uint8Array(32).fill(index + 1) }));
assert.equal((await decodeNativeCapabilitySet(await encodeNativeCapabilitySet(views))).length, 32);
await assert.rejects(encodeNativeCapabilitySet([...views, { ...view, account: new Uint8Array(32).fill(33) }]));
for (const grant of all) {
  if (grant.kind === "program_spend") {
    const sourceAccount = new Uint8Array(grant.sourceAccount); sourceAccount[0] = sourceAccount[0]! ^ 1;
    await assert.rejects(encodeNativeCapabilitySet([{ ...grant, sourceAccount }]));
    await assert.rejects(encodeNativeCapabilitySet([{ ...grant, seed: new Uint8Array(129) }]));
  }
  if (grant.kind === "transfer402") await assert.rejects(encodeNativeCapabilitySet([{ ...grant, maximumAmount: 0n }]));
  if (grant.kind === "balance_view") await assert.rejects(encodeNativeCapabilitySet([{ ...grant, receiptDigest: new Uint8Array(32) }]));
}
